planner/core/hints.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/cznic/mathutil"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb/sessionctx"
+)
+
+// BlockHintProcessor processes hints at different level of sql statement.
+type BlockHintProcessor struct {
+	QbNameMap        map[string]int                    // Map from query block name to select stmt offset.
+	QbHints          map[int][]*ast.TableOptimizerHint // Group all hints at same query block.
+	Ctx              sessionctx.Context
+	selectStmtOffset int
+}
+
+// Enter implements Visitor interface.
+func (p *BlockHintProcessor) Enter(in ast.Node) (ast.Node, bool) {
+	switch node := in.(type) {
+	case *ast.UpdateStmt:
+		p.checkQueryBlockHints(node.TableHints, 0)
+	case *ast.DeleteStmt:
+		p.checkQueryBlockHints(node.TableHints, 0)
+	case *ast.SelectStmt:
+		p.selectStmtOffset++
+		node.QueryBlockOffset = p.selectStmtOffset
```

Comment:
Shall we start the naming from `sel_0` or `sel_1`?

Reply:
`sel_1`, always number from 1.